src/mesh/mesh_adaptation/anisotropic_mesh_adaptation.cpp

```diff
@@ -0,0 +1,483 @@
+    , mpi_communicator(MPI_COMM_WORLD)
+    , pcout(std::cout, dealii::Utilities::MPI::this_mpi_process(mpi_communicator)==0)
+    , initial_poly_degree(dg->get_min_fe_degree())
+    , max_dofs_per_cell(dg->dof_handler.get_fe_collection().max_dofs_per_cell())
+{
+    MPI_Comm_rank(mpi_communicator, &mpi_rank);
+    MPI_Comm_size(mpi_communicator, &n_mpi);
+    
+    if(use_goal_oriented_approach)
+    {
+        if(normLp != 1)
+        {
+            std::cout<<"Optimal metric for the goal oriented approach has been derived w.r.t the error in L1 norm. Aborting..."<<std::flush;
+            std::abort();
+        }
+        
+        std::shared_ptr<Physics::ModelBase<dim,nstate,real>> pde_model_double    = Physics::ModelFactory<dim,nstate,real>::create_Model(dg->all_parameters);
+        pde_physics_double  = Physics::PhysicsFactory<dim,nstate,real>::create_Physics(dg->all_parameters, pde_model_double);
+        functional = FunctionalFactory<dim,nstate,real,MeshType>::create_Functional(dg->all_parameters->functional_param, dg);
+    }
+
+    if(dg->get_min_fe_degree() != dg->get_max_fe_degree())
+    {
+        std::cout<<"This class is currently coded assuming a constant poly degree. To be changed in future if required."<<std::endl;
+        std::abort();
+    }
+
+    if(initial_poly_degree != 1)
+    {
+        pcout<<"Warning: The optimal metric used by this class has been derived for p1."
+             <<" For any other p, it might be a good approximation but will not be optimal."<<std::endl; 
+    }
+
+    Assert(this->dg->triangulation->get_mesh_smoothing() == typename dealii::Triangulation<dim>::MeshSmoothing(dealii::Triangulation<dim>::none),
+           dealii::ExcMessage("Mesh smoothing might h-refine cells while changing p order."));
+
+}
+
+template<int dim, int nstate, typename real, typename MeshType>
+dealii::Tensor<2, dim, real> AnisotropicMeshAdaptation<dim, nstate, real, MeshType> 
+    :: get_positive_definite_tensor(const dealii::Tensor<2, dim, real> &input_tensor) const
+{
+    // Check if the tensor is symmetric.
+    dealii::Tensor<2, dim, real> input_tensor_copy = input_tensor;
+    for(int i=0; i<dim; ++i)
+    {
+        for(int j=0; j<dim-i; ++j)
+        {
+            if(abs(input_tensor_copy[i][j] - input_tensor_copy[j][i]) > 1.0e-10)
+            {
+                std::cout<<"Input tensor needs to be symmetric. Difference = "<< abs(input_tensor_copy[i][j] - input_tensor_copy[j][i]) <<".  Aborting.."<<std::endl<<std::flush;
+                std::abort();
+            }
+            else 
+            {
+                input_tensor_copy[j][i] = input_tensor_copy[i][j];
+            }
+        }
+    }
+    dealii::SymmetricTensor<2,dim,real> symmetric_input_tensor(input_tensor_copy); 
+    std::array<std::pair<real, dealii::Tensor<1, dim, real>>, dim> eigen_pair = dealii::eigenvectors(symmetric_input_tensor);
+
+    std::array<real, dim> abs_eigenvalues;
+    const real min_eigenvalue = 1.0e-8;
```

Comment:
These limits are basically user defined tolerances set to prevent degenerate cases (abs(eigenvalues) becoming zero or infinity) and, while the upper bound might be difficult to achieve, it is possible for the eigenvalue to be close to zero (for example, Hessian can have zero eigenvalues when the exact solution of the PDE is nearly linear).

Reply:
Sounds good.